lemur/plugins/lemur_kubernetes/plugin.py

```diff
@@ -70,92 +118,179 @@ class KubernetesDestinationPlugin(DestinationPlugin):
     ]
 
     def __init__(self, *args, **kwargs):
         super(KubernetesDestinationPlugin, self).__init__(*args, **kwargs)
 
     def upload(self, name, body, private_key, cert_chain, options, **kwargs):
 
-        k8_bearer = self.get_option('kubernetesAuthToken', options)
-        k8_cert = self.get_option('kubernetesServerCertificate', options)
-        k8_namespace = self.get_option('kubernetesNamespace', options)
-        k8_base_uri = self.get_option('kubernetesURL', options)
-
-        k8s_api = K8sSession(k8_bearer, k8_cert)
-
-        cert = Certificate(body=body)
-
-        # in the future once runtime properties can be passed-in - use passed-in secret name
-        secret_name = 'certs-' + urllib.quote_plus(cert.name)
-
-        err = ensure_resource(k8s_api, k8s_base_uri=k8_base_uri, namespace=k8_namespace, kind="secret", name=secret_name, data={
-            'apiVersion': 'v1',
-            'kind': 'Secret',
-            'metadata': {
-                'name': secret_name,
-            },
-            'data': {
-                'combined.pem': base64.b64encode(body + private_key),
-                'ca.crt': base64.b64encode(cert_chain),
-                'service.key': base64.b64encode(private_key),
-                'service.crt': base64.b64encode(body),
-            }
-        })
+        try:
+            k8_base_uri = self.get_option('kubernetesURL', options)
+            secret_format = self.get_option('secretFormat', options)
+
+            k8s_api = K8sSession(
+                self.k8s_bearer(options),
+                self.k8s_cert(options)
+            )
+
+            cn = common_name(parse_certificate(body))
+
+            secret_name_format = self.get_option('secretNameFormat', options)
+
+            secret_name = secret_name_format.format(common_name=cn)
+
+            secret = build_secret(secret_format, secret_name, body, private_key, cert_chain)
+
+            err = ensure_resource(
+                k8s_api,
+                k8s_base_uri=k8_base_uri,
+                namespace=self.k8s_namespace(options),
+                kind="secret",
+                name=secret_name,
+                data=secret
+            )
+
+        except Exception as e:
+            current_app.logger.exception("Exception in upload")
+            raise e
 
         if err is not None:
+            current_app.logger.debug("Error deploying resource: %s", err)
             raise Exception("Error uploading secret: " + err)
 
+    def k8s_bearer(self, options):
+        bearer = self.get_option('kubernetesAuthToken', options)
+        if not bearer:
+            bearer_file = self.get_option('kubernetesAuthTokenFile', options)
+            with open(bearer_file, "r") as file:
+                bearer = file.readline()
+            if bearer:
+                current_app.logger.debug("Using token read from %s", bearer_file)
+            else:
+                raise Exception("Unable to locate token in options or from %s", bearer_file)
```

Comment:
Should we log if this exception is raised?

Reply:
That exception gets logged in the exception handler in upload. Would you like me to add a log entry here?